### ISSUE TITLE
Graph definitions are not refreshed for repeating panels/rows

### ISSUE THREAD

**Issue by RichiH** [grafana/grafana#3751]
Using the Prometheus data source with templates which are set to refresh every time the dashboard is openend.

If I edit my graph definition, e.g. by flipping the Y axis to negative for some lines, and saving, this change is not copied to the repeating panels.

What works for repeating panels is to:

1. Disable Repeat Panel
2. Click save
3. Enable Repeat Panel

for repeating rows it's even worse:

1. Disable Repeat Row
1. Click save
1. Full reload of the dashboard with CTRL-F5
1. Enable Repeat Row
1. Click save

Relatedly, I also sometimes get dashboards with repeated panels/row which show "No Data Points"; doing the above fixes that.

Long story short, I suspect Grafana caches / keeps state about some of those graphs and does not properly invalidate old items.

**Comment by torkelo** [comment]
you use editing the source panel/row (the one being repeated) and not the clones? Make sure you edit the source panel. Then to trigger a repeat cycle you need to change a template variable (or save/reload), this should remove all the cloned panels/rows and re-repeat them

**Comment by torkelo** [comment]
closing for lack of update

**Comment by RichiH** [comment]
Ugh, sorry, I thought I had replied.

Just tested with current Grafana from official RPMs; this bug is still valid.

* Change source panel:
`irate(ifHCInOctets{fqdn=~"$ciscos",ifDescr=~"$cisco_interfaces"}[2m])*8` to `irate(ifHCInOctets{fqdn=~"$ciscos",ifDescr=~"$cisco_interfaces"}[2m])*8000`
* Click "Back to dashboard"
* Grafana starts reloading all panels
  * The first, source panel shows a thousand-fold increase
  * All other panels reload, but do *NOT* show the thousand-fold increase
* Click "Save Dashboard"
* Once Grafana is done loading, it shows "Dashboard saved"

At this point, only the source panel shows the changed graph

* Click Reload in the browser navigation bar / press F5

Only now, the other panels show the change.

**Comment by torkelo** [comment]
this is not a bug, if you change the source panel you need to trigger a revaluation of the panel repeats. This is done by either changing the variable value that is the basis for the panel repeats, or reloading the dashboard.

**Comment by RichiH** [comment]
It does **not** work when you're changing the variable value that is the basis for the panel repeats.

It only works when you reload.

**Comment by torkelo** [comment]
when ever the server variable, the panel repeat is happening.
http://play.grafana.org/dashboard/db/graphite-panel-repeats

**Comment by RichiH** [comment]
Can you suggest an edit which I can play with? I don't know Graphite at all; and the Prometheus dashboard does not have templates I can repeat on.

I suspect that this problem might be specific to Prometheus data sources.

**Comment by RichiH** [comment]
I lack the permissions to save the live Prometheus dashboard (of course), and I can't get it to repeat on a template without reloading.

So, I would say that play.grafana.org shows the same bug.

**Comment by torkelo** [comment]
I am not sure how this can be a prometheus related issue. 

If you change the template variable that is used to repeat a panel. changes to the first (source) panel should be propagated to the repeated panels. The same code is used for that no matter what data source.

**Comment by RichiH** [comment]
I agree that that should happen, but I can tell you that it does not happen. ;)

Should I send you screenshots of the effect?

**Comment by RichiH** [comment]
Just to make that part clear: Grafana _does_ reload all panels. The swirly "please wait" thing in the upper right circles for some time and then the graphs are redrawn. But the panel does not actually change, taking the new query into account.

As an aside, please re-open this issue so we don't lose track of it by mistake,

**Comment by torkelo** [comment]
@RichiH here is a prometheus dashboard with repeats
http://play.grafana.org/dashboard/db/prometheus-repeat

When I change the stack option for example on the first graph, then change the job variable, the change will propagate

**Comment by torkelo** [comment]
just tested updating the query on the first panel, and it is propagated to the other panels after I change the job variable.

**Comment by RichiH** [comment]
@torkelo I can reproduce in your demo panel:

* Click Disk I/O title
* Click edit
* Change `go_memstats_alloc_bytes` to `go_memstats_alloc_bytes/100000`
* Click "Back to dashboard"
* See this effect.

**Comment by RichiH** [comment]
Using Chrome Version 51.0.2704.63 (64-bit) on Debian unstable.

**Comment by torkelo** [comment]
You are missing the step that I have repeated 10 times, you need to change the variable that you repeat by

**Comment by torkelo** [comment]
so after you go back to dashboard, update the variable (just change one selected option), then you can change it back. Not ideal I know but it's currently the only way to trigger a update that rebuilds the repeated dashboards. Since this is during dashboard building phase it should be fine.

**Comment by RichiH** [comment]
Ah yes, but that does changes nothing:

* Click Disk I/O title
* Click edit
* Change `go_memstats_alloc_bytes` to `go_memstats_alloc_bytes/100000`
* Click "General"
* Change "Repeat Panel" to empty
* Click "Back to dashboard"
* Click Refresh _in the Grafana page_
* See this effect.

* Click Disk I/O title
* Click edit
* Click "General"
* Change "Repeat Panel" to "job"
* Click "Back to dashboard"
* Click Refresh _in the Grafana page_
* Same effect

**Comment by RichiH** [comment]
Ah, you mean "change it in the navigation", not "change it in edit menu".

Yes, that works.

I would _strongly_ argue that you should display a warning when changing repeat panels as long as this behaviour can not be improved.

**Comment by torkelo** [comment]
@RichiH I am must be expressing my self badly because you still do not understand what I am saying. I did not say refresh, I said CHANGE THE VALUE OF THE FUCKING VARIABLE! 

sorry for the caps :)  you need to change the selected options in the variable for the variable to trigger a dynamic dashboard update (panel repeats are deleted/added/updated)

**Comment by RichiH** [comment]
Yes, I got that now :)

Now that I know this, it's reasonably obvious **in hindsight** and I won't forget; problem is that it's not intuitive and thus confusing for people who don't happen to be developers of Grafana ;)

**Comment by RichiH** [comment]
More to the point, a refresh is intuitively stronger/more invasive than changing a single variable in a page.

**Comment by torkelo** [comment]
I understand that, and agree it could be more intuitive. But so far no one has had any issues with. But I agree it could be better.  

Doing it on refresh is very expensive, as it would involve a dirty check and a dynamic dashboard reevaluation (delete/add/update repeat panels), very expensive, both the detect change and rebuild phase. that is why it is only done when the repeat variable is changed.

Will try to add more docs on this when I have time

**Comment by torkelo** [comment]
and please read what I write more closely in the future :)
In the very first comment I wrote:
> Then to trigger a repeat cycle you need to change a template variable (or save/reload), this should remove all the cloned panels/rows and re-repeat them

**Comment by RichiH** [comment]
Having talked to others: They saw it, they just didn't report it.

The simple way out would be to show an info bubble in the upper right every time the "Repeat Panel" is changed. That way, everyone creating repeats sees this info.

**Comment by RichiH** [comment]
I did read it, but I thought you meant "change the value in 'Repeat Panel'" - which I did, and which does nothing. Now that I think about it, you might almost call that... a bug ;)

**Comment by RichiH** [comment]
But yes, I understand your frustration as you thought you had communicated clearly and I was just too dense.

**Comment by torkelo** [comment]
please send PR when anything is changed in a panel that shows this info, not very easy to do in a elegant way.  there is a reason why it's this way. But hope to improve the docs for repeat panel. Repeat panels is an advanced feature is currently undocumented. Sadly no one has helped improving this feature.

**Comment by torkelo** [comment]
argh, sorry for being frustrated. just is tiring to respond to 100 github issues every day and say much the same thing so often, sometimes it results in frustration that is unwarranted :)

**Comment by torkelo** [comment]
I dont want you to hesitate to open issues again, bug reports and improvement suggestions are super important :)

**Comment by RichiH** [comment]
Yah, no worries; I won't go away ;)

I might be able to look at this over the weekend, but I do not speak Go, so...

Should we re-open this issue so the info part is not forgotten?

**Comment by forsberg** [comment]
For the record, this confused me as well, and I had to read through this whole task until I understood how to trigger repeat. It did not help that I had hidden the template variable ('host' in my case).

So, for all the others, if you have a repeating panel, and it doesn't update:

0. Do not hide your template variable.
1. Do the change, and do it on the top row (which I believe is always the source row).
2. Save
3. Change what's selected in the dropdown for your template variable, for example from "all" to one single value. Then change back to "all". 
4. Dashboard is rebuilt.

**Comment by relishcolouredhat** [comment]
To anybody else who found themselves here, trying to rebuild their dash filled with hidden template variables (basically just paraphrasing above answer, which was incomplete):

1. Make desired change to repeated template panel, or panel(s) in a repeated row;
2. Unhide variable repeated panel/row is based upon if it is hidden;
3. Select a **single** item within the dropdown menu of the variable the template is based upon - you must cause the 'template generated panels' to *go away* for them to be regenerated
4. Modify dropdown menu selection to the desired state, causing the repeated panel/row to be rebuilt-- with your updated changes.

Apologies to re-open a dead thread, but I spent more time than I care to admit trying to figure out the bare minimum to trigger this change - no save is actually required, just the 're-generation' of the repeated portion.

Probably not the place to mention this, but having some function to reload this without having to clear the selections would be really useful for my environment - I use hidden template variables quite a bit, and have some use cases where 10-20 streams from a list of hundreds are selected once and never changed.

Since UI interaction is required it my current method is to capture variables from url with 'share' functionality, manually deselect templated streams, save, and then pass the variables back in via url. Not that bad to deal with, but would be super nice to just have some sort of 'reload all' button that didn't require clearing and re-feeding the variables.

**Comment by shirakaba** [comment]
As a TL;DR to others ending up here because edits to repeated panels are not being committed even upon save:

> are you editing the source panel/row (the one being repeated) and not the clones? Make sure you edit the source panel.

For me, that meant that I should edit the left-most panel in a repeating set rather than the right-most one (otherwise my changes would be lost upon altering the displayed template variables).
